### PR TITLE
feat(test): add docs linter to PR workflow actions

### LINKED ISSUE #4320: feat: add linters for Docs as part of PR testing pipeline
**Is your feature request related to a problem? Please describe.**
Yes: Learning and implementing NetlifyCMS with confidence is very difficult because the Docs lack completeness due to an inconsistent and subjectively validated contribution workflow.
(Please see #4281 for in-depth explanations)

**Describe the solution you'd like**
Implement automated validation (linting) for the Docs as part of the PR testing pipeline as a first step in improving the contribution workflow and ensuring the content stays updated and accurate.

**Two insertion points for Docs validation:**
1. Add prose linting for Docs as part of local tests

2. Add two GitHub Checks to the 12 existing PR checks:
```
13. Docs code samples are correct: pass/fail - link to detailed linter output
14. Docs prose style & usage is correct: pass/fail - link to detailed linter output
```

**Describe alternatives you've considered**
It would be ideal if prose linting and code example linting happened live in the CMS editor... but that is probably a more complex implementation.

**Additional context**
[20 min video about real-time prose linting by one of the creators of ProseLint](https://www.youtube.com/watch?v=S55EFUOu4O0)

### PR DESCRIPTION
**Pull request by silverwebb925**
Closes #4320 

fizzgig_mcfloofen
> Fizzgig McFlööfen w/ all the toys

### CODE REVIEW

**Comment by silverwebb925** [comment]
The Docs Linting test worked on the first commit, but not the subsequent... closing and opening a different PR to see what happens... apologies for the repeat.

**Comment by silverwebb925** [comment]
**Good news**: Vale prose linter incorporated as a GitHub Action, with initial basic prose linting configured as part of the `.vale` directory. The linting isn't very useful right now, but is highly extendable and easy to work with. Assuming it fits the bill in the pipeline, the rules themselves are easy enough to formulate.
Configured via the `.github/workflows/docslint.yml` file.

**Bad news**: to the best of my understanding, the GitHub Actions `pull_request` event is passed to the base repo, not the forked repo... which means it would only lint the existing documentation, not anything submitted as part of a PR... right? _Please confirm! (@erezrokah ?)_

Currently, I reconfigured the GH action to trigger on a `push` event because I wanted to test the configuration.

**References:**
1. [GitHub: Events that trigger workflows](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request)
2. [Vale Linter](https://docs.errata.ai/vale/about)
3. [Vale Action](https://github.com/errata-ai/vale-action)

**Next Steps:**
Vale is definitely the right track to get documentation linting into the CI/CD pipeline, but I'm pushing the envelope of my dev skills as it is and need input about next steps.

- **Check out the workflow output on the [Actions page](https://github.com/silverwebb925/netlify-cms/actions) of my forked repo.**

**Comment by erezrokah** [comment]
Thanks for following up on this @silverwebb925.

> **Bad news**: to the best of my understanding, the GitHub Actions `pull_request` event is passed to the base repo, not the forked repo... which means it would only lint the existing documentation, not anything submitted as part of a PR... right? _Please confirm! (@erezrokah ?)_

Not sure I understand the issue. `pull_request` is passed, we used it in another action:
https://github.com/netlify/netlify-cms/blob/db86ea295474ed185fa42b3144373eec93a08221/.github/workflows/nodejs.yml#L105

Looks like the action is running `vale` locally:
https://github.com/errata-ai/vale-action/blob/75a4db25a0833de205ab750af4b4ed36e24280ec/src/main.ts#L19
so that should work for forked repos too.

The other issue I was concerned about was that PRs running from forks have limited scopes for the GitHub token and can't write status checks:
https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token

however it seems they handle that use case:
https://github.com/errata-ai/vale-action/blob/75a4db25a0833de205ab750af4b4ed36e24280ec/src/check.ts#L103

The action log seems ok to me:
https://github.com/netlify/netlify-cms/pull/4321/checks?check_run_id=1120652708#step:4:8

Is there a way to extend some of the settings in the PR? I'm not sure we would want to maintain all those files.

**Comment by silverwebb925** [comment]
> Not sure I understand the issue. `pull_request` is passed, we used it in another action:
You're absolutely right! I realized that the next morning after my GoodNews/BadNews comment and updated the action to match the NodeJS test... just forgot to update my comment (sorry!).

> The other issue I was concerned about was that PRs running from forks have limited scopes for the GitHub token and [can't write status checks](https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token)
Again, right on the mark. I wasn't able to articulate this (nor did I have the specific reference). Thanks for confirming this is the case... it's a bummer and I hope at some point the annotations can be extended to Forks... but for now I agree that the current action log seems okay.
> https://github.com/netlify/netlify-cms/pull/4321/checks?check_run_id=1120652708#step:4:8

> Is there a way to extend some of the settings in the PR? I'm not sure we would want to maintain all those files.
The short answer is "Yes, absolutely".
Are you concerned about the idea of having hundred's of warnings/errors every PR? Because that is _definitely_ not the goal here... but I think it would be good to check all the markdown files for the website during every PR so that **_any_** content updates get caught and linted for style and usage.

As for the current warning/errors... dialing in the style & usage rules should reduce the number of errors significantly. For example:
Without any configuration:
`Vale: Found 0 suggestion(s), 0 warning(s), and 1126 error(s).`
With my zero-pass at rule configuration:
`Vale: Found 0 suggestion(s), 371 warning(s), and 30 error(s).`

As the linter rules get dialed in to match NetlifyCMS style & usage guidelines, we can reduce the errors/warnings to zero for all existing content through revision and resubmission.

Alternatively, I suppose we could change the config now so that only changed files are linted... and then work through the Docs as things come up?

**Comment by erezrokah** [comment]
> Are you concerned about the idea of having hundred's of warnings/errors every PR? Because that is _definitely_ not the goal here... but I think it would be good to check all the markdown files for the website during every PR so that **_any_** content updates get caught and linted for style and usage.

I'm concerned about maintaining all those different configurations, especially when most maintainers lack the knowledge to do so.
Is there a way to reduce the amount of errors/warnings without them?

> Alternatively, I suppose we could change the config now so that only changed files are linted... and then work through the Docs as things come up?

I like this idea. This way we can gradually introduce this tool.

**Comment by silverwebb925** [comment]
> > Alternatively, I suppose we could change the config now so that only changed files are linted... and then work through the Docs as things come up?
> I like this idea. This way we can gradually introduce this tool.

Actually, I thought through this more, and until we can incorporate linter-review as part of the Open Authoring workflows, I think we should keep the prose linter checking all the docs files for now.

> I'm concerned about maintaining all those different configurations, especially when most maintainers lack the knowledge to do so.
> Is there a way to reduce the amount of errors/warnings without them?

Okay, so I spent today dialing in the prose linter with a really solid first pass at rule making and consolidation.

Netlify CMS has a really solid generic prose linter now, _and_ I hooked up the "avoid using this word/phrase" rules as a `file collection` that is accessible through the CMS. As it is, a non-technical user can now add/remove words from the rulesets, but some additional work will be required to make it more accessible (the csv default for the list widget isn't particularly useful).

Also, I started reviewing the warnings and errors to correct the docs.
I didn't do a full analysis, but here's a few interesting numbers:
76.8% of the warnings relate to the use of passive voice
13.5% of the warnings relate to the use of adverbs
This accounts for 90% of the errors!

We'll have to work through the individual articles, but once we clear these 333 (now 314), we should have a relatively low warning count (only as new documentation is added).

**Comment by erezrokah** [comment]
Thank you again @silverwebb925 for all this work and guiding me through the process.
Everything you said makes sense and hooking the rules to a file widget is a great bonus.
Will experiment with this a bit more to make sure I understand everything that is going on.

**Comment by silverwebb925** [comment]
> Thank you again @silverwebb925 for all this work and guiding me through the process.
> Everything you said makes sense and hooking the rules to a file widget is a great bonus.
> Will experiment with this a bit more to make sure I understand everything that is going on.

You're most welcome. This is good experience for me as I learn software dev, and a subject I'm passionate about. Not to mention, that having good documentation means I get to learn and implement Netlify CMS easier for other projects I'm working on.

Let me know if anything seems confusing about the setup right now. This feels like a good initial implementation, but it won't be much good as a writing tool until it can be incorporated with the Open Authoring process. I'm not sure _this_ PR is quite ready to merge, but it's a lot closer after my work yesterday.
At a miminum:
- [ ] Incorporate the existing `writing-guide.md` styles as linter rules.

It's worth noting that I temporarily moved the `writing-guide.md` out of the docs pages and into `website/src/writing-guide`. I did this for two reasons:
1. Almost all the "Don't" examples in the guide were getting flagged by the linter.
2. Ideally, the docs authoring guide would be accessible in the CMS instead of existing as an external-facing guide.
    Even better, if there was a way to have a quick-reference guide in the editor itself for basic styles.

The external-facing contributor guide should still exist of course, but focus on contributing code, with a short blurb explaining how to log into the open authoring platform and access the authoring guide.

Additional work:
- [ ] Update list widget to better handle CSV lists... parameters for formatting the list would be nice.
- [ ] Creating a table widget would be better. Essentially a list-widget with the ability to resort items (to begin with).
- [ ] Get the code-linters running on the code blocks in the docs...
- [ ] Figure out how to flagged out-of-date articles for review and revision.

File: .vale.ini
Comment (erezrokah) [inline .vale.ini:2]: This didn't work for me locally after installing `brew install vale`
I suspect it is because the action docker image uses an older version of `vale`.
The docs for `Vocab` specify a different structure than in this PR

File: website/src/writing-guide/styles/Replace/Offensive.yml
Comment (erezrokah) [inline website/src/writing-guide/styles/Replace/Offensive.yml:8]: Had to change these to a string `[CENSORED]->CENSORED` for it to run locally. Probably due to versions gap.

**Review by erezrokah — APPROVED** [review]

**Comment by erezrokah** [comment]
I'm good with merging this as it is much better than the current state. This sets us in a great place to incrementally improve our docs.

File: website/src/writing-guide/styles/Replace/Offensive.yml
Comment (silverwebb925) [reply website/src/writing-guide/styles/Replace/Offensive.yml:8]: That makes sense. Thank you for updating it.

File: .vale.ini
Comment (silverwebb925) [reply .vale.ini:2]: I haven't evaluated the local version at all, so can't speak to this difference... I'm curious about running it locally, but  the goal is to eventually incorporate the documentation linting into the editor itself so that authors can have a purely CMS interaction.

**Comment by silverwebb925** [comment]
Thank you for finalizing and merging this PR, @erezrokah.
I'll circle back soon regarding next steps and to increment further.
Appreciate your attention and feedback on my work, as well as the work you're putting in yourself.

**Pull request merged**
